### PR TITLE
Fix ecred connection retry on encryption change

### PR DESCRIPTION
**Pull request by hermabe**
The retry mechanism on encryption change after rejection due to insufficient encryption level did not work. The L2CAP channels were removed from the `bt_conn` channels list. They were thus not found in the ecred connection response handler.

The test from the L2CAP collision mitigation PR is used to verify the change, with `CONFIG_BT_EATT_SEC_LEVEL=2` instead of the default level 1.

### CODE REVIEW

**Review by jhedberg — DISMISSED** [review]

**Comment by alwa-nordic** [comment]
Could you help me understand what function `l2cap_le_encrypt_change` is intended to perform? (I miss descriptive function names. ~) I am especially confused about why this function is invoked for every chan in a conn, and then this function again loops over every chan in the conn of the chan provided.

**Comment by hermabe** [comment]
When a ECRED connection response is received and the request has failed because of authentication, the host tries to enable encryption and retry the connection. `bt_conn_set_security()` will send an HCI command to set up encryption, and the `BT_L2CAP_STATUS_ENCRYPT_PENDING` flag is set on each channel. `bt_l2cap_security_changed()` is called when encryption setup is finished. For the channels that are pending encryption a new connection request is sent. The channels with the same `ident` belong to the same connection request and are grouped together for the retry request.

I see now why the channels were removed from the `conn->channels` list. As you pointed out, `l2cap_le_encrypt_change()` is called per channel, and the host should not initiate a new connection request for each channel, as they are grouped together. I think the `BT_L2CAP_STATUS_ENCRYPT_PENDING` flag should be cleared on all the channels that are grouped in the retry request. Then the calls to `l2cap_le_encrypt_change()` will have no effect for them.

**Review by jhedberg — DISMISSED** [review]

**Comment by hermabe** [comment]
The CI failure seems to exist on main as well, so will have to rebase once that is fixed.

File: subsys/bluetooth/host/l2cap.c [hunk 1]
```diff
@@ -551,11 +551,14 @@ static void l2cap_le_encrypt_change(struct bt_l2cap_chan *chan, uint8_t status)
...
 	if (chan->ident) {
 		struct bt_l2cap_chan *echan[L2CAP_ECRED_CHAN_MAX];
-		struct bt_l2cap_le_chan *ch;
+		struct bt_l2cap_chan *ch;
 		int i = 0;
 
-		while ((ch = l2cap_remove_ident(chan->conn, chan->ident))) {
-			echan[i++] = &ch->chan;
+		SYS_SLIST_FOR_EACH_CONTAINER(&chan->conn->channels, ch, node) {
+			atomic_clear_bit(ch->status, BT_L2CAP_STATUS_ENCRYPT_PENDING);
```
Comment (alwa-nordic) [inline subsys/bluetooth/host/l2cap.c:558]: We should only clear the `BT_L2CAP_STATUS_ENCRYPT_PENDING` flag when the `indent` matches and the `ch` is added to `echan`.

File: subsys/bluetooth/host/l2cap.c [hunk 2]
```diff
@@ -551,11 +551,14 @@ static void l2cap_le_encrypt_change(struct bt_l2cap_chan *chan, uint8_t status)
...
+			if (chan->ident == ch->ident) {
+				echan[i++] = ch;
```
Comment (alwa-nordic) [inline subsys/bluetooth/host/l2cap.c:560]: I think static analysis will infer a possible-out-of-bounds-write. We should add `if(i == 5) { break; }` or assert `i < 5`.

**Review by alwa-nordic — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/host/l2cap.c (hunk 1 above)
Comment (hermabe) [reply subsys/bluetooth/host/l2cap.c:558]: Yes, that is the wrong place.

File: subsys/bluetooth/host/l2cap.c (hunk 2 above)
Comment (hermabe) [reply subsys/bluetooth/host/l2cap.c:560]: There should never be more than 5 channels with the same `ident`, so will add an assert.

**Review by Thalley — APPROVED** [review]

**Review by alwa-nordic — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Pull request merged**